app/code/community/Bolt/Boltpay/Helper/CatalogHelper.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+class Bolt_Boltpay_Helper_CatalogHelper extends Mage_Core_Helper_Abstract
+{
+    /**
+     * @return Mage_Core_Model_Abstract
+     */
+    public function getSession()
+    {
+        return Mage::getSingleton('catalog/session');
+    }
+
+    public function getLastRealOrderId()
+    {
+        return Mage::getSingleton('catalog/session')->getLastRealOrderId();
+    }
+
+    /**
+     * @param string $scope
+     * @return string
+     */
+    private function getQuoteIdKey($scope = 'ppc')
+    {
+        return $scope.'_quote_id_' . Mage::app()->getStore()->getId();
+    }
+
+    /**
+     * Get Quote for Product page
+     * @return Mage_Sales_Model_Quote
+     */
+    public  function getQuote()
+    {
+        $hasOrder = false;
+        $orderQuoteId = false;
+        $ppcQuoteId = $this->getSession()->getData($this->getQuoteIdKey());
+        if ($ppcQuoteId) {
+            $order = Mage::getModel('sales/order')->getCollection()->addAttributeToFilter('entity_id', $this->getLastRealOrderId())->getFirstItem()->setPageSize(1);
+            if ($order instanceof Mage_Sales_Model_Order) {
+                $orderQuoteId = $order->getQuoteId();
+            }
+            /** @var Mage_Sales_Model_Quote $ppcQuote */
+            $ppcQuote = Mage::getModel('sales/quote')->loadByIdWithoutStore($ppcQuoteId);
+            if ($orderQuoteId && $ppcQuoteId == $orderQuoteId) {
+                $ppcQuote->setIsActive(false);
+                $ppcQuote->delete();
+                $hasOrder = true;
+            }
+        }
+        if ($hasOrder || empty($ppcQuote)) {
+            /** @var Mage_Sales_Model_Quote $ppcQuote */
+            $ppcQuote = Mage::getModel('sales/quote');
+            $ppcQuote->setStore(Mage::app()->getStore());
+            $ppcQuote->reserveOrderId();
+        }
+        return $ppcQuote;
+    }
+
+    /**
+     * @return Mage_Sales_Model_Quote
+     */
+    public function getQuoteWithCurrentProduct(Mage_Catalog_Model_Product $product, $request = null)
+    {
+        /** @var Mage_Sales_Model_Quote $ppcQuote */
+        $ppcQuote = $this->getQuote();
+        $ppcQuote->removeAllItems();
+        $ppcQuote->addProduct($product, $request);
+        $ppcQuote->setParentQuoteId($ppcQuote->getId());
+        $ppcQuote->getShippingAddress()->setCollectShippingRates(true);
```

Comment:
Please add `$ppcQuote->getBillingAddress()`,  this method add proper quote reference to the billing address.

Reply:
The Magento do it in `\Mage_Checkout_Model_Cart::save()` :
```
/**
     * Save cart
     *
     * @return Mage_Checkout_Model_Cart
     */
    public function save()
    {
        Mage::dispatchEvent('checkout_cart_save_before', array('cart'=>$this));

        $this->getQuote()->getBillingAddress();
        $this->getQuote()->getShippingAddress()->setCollectShippingRates(true);
        $this->getQuote()->collectTotals();
        $this->getQuote()->save();
        $this->getCheckoutSession()->setQuoteId($this->getQuote()->getId());
        /**
         * Cart save usually called after changes with cart items.
         */
        Mage::dispatchEvent('checkout_cart_save_after', array('cart'=>$this));
        return $this;
    }
```